src/document/doc-utils.js

```diff
@@ -180,42 +180,135 @@ function removeLines(doc) {
   return mapDoc(doc, removeLinesFn);
 }
 
-function getInnerParts(doc) {
-  let { parts } = doc;
-  let lastPart;
-  // Avoid a falsy element like ""
-  for (let i = doc.parts.length; i > 0 && !lastPart; i--) {
-    lastPart = parts[i - 1];
+const isHardline = (doc, nextDoc) =>
+  doc &&
+  doc.type === "line" &&
+  doc.hard &&
+  nextDoc &&
+  nextDoc.type === "break-parent";
+function stripDocTrailingHardlineFromDoc(doc) {
+  if (!doc) {
+    return doc;
   }
-  if (lastPart.type === "group") {
-    parts = lastPart.contents.parts;
+
+  switch (doc.type) {
+    case "concat":
+    case "fill": {
+      const { parts } = doc;
+
+      while (parts.length > 1 && isHardline(...parts.slice(-2))) {
+        parts.length -= 2;
+      }
+
+      if (parts.length > 0) {
+        const lastPart = stripDocTrailingHardlineFromDoc(
+          parts[parts.length - 1]
+        );
+        parts[parts.length - 1] = lastPart;
+      }
+      break;
+    }
+    case "align":
+    case "indent":
+    case "group":
+    case "line-suffix":
+      doc.contents = stripDocTrailingHardlineFromDoc(doc.contents);
+      break;
+    case "if-break":
+      doc.breakContents = stripDocTrailingHardlineFromDoc(doc.breakContents);
+      doc.flatContents = stripDocTrailingHardlineFromDoc(doc.flatContents);
+      break;
   }
-  return parts;
+
+  return doc;
 }
 
-function stripTrailingHardline(doc, withInnerParts = false) {
+function stripTrailingHardline(doc) {
   // HACK remove ending hardline, original PR: #1984
-  if (doc.type === "concat" && doc.parts.length !== 0) {
-    const parts = withInnerParts ? getInnerParts(doc) : doc.parts;
-    const lastPart = parts[parts.length - 1];
-    if (lastPart.type === "concat") {
+  return stripDocTrailingHardlineFromDoc(cleanDoc(doc));
+}
+
+const isConcat = (doc) => doc && doc.type === "concat";
+function cleanDocFn(doc) {
+  switch (doc.type) {
+    case "fill":
+      if (doc.parts.length === 0 || doc.parts.every((part) => part === "")) {
+        return "";
+      }
+      if (doc.parts.length === 1) {
+        return doc.parts[0];
+      }
+      break;
+    case "group":
+      // Remove nested only group
       if (
-        lastPart.parts.length === 2 &&
-        lastPart.parts[0].hard &&
-        lastPart.parts[1].type === "break-parent"
+        doc.contents.type === "group" &&
+        doc.contents.id === doc.id &&
+        doc.contents.break === doc.break &&
+        doc.contents.expandedStates === doc.expandedStates
       ) {
-        return { type: "concat", parts: parts.slice(0, -1) };
+        return doc.contents;
+      }
+      break;
+    case "align":
+    case "indent":
+    case "line-suffix":
+      if (!doc.contents) {
+        return "";
       }
+      break;
+    case "if-break":
+      if (!doc.flatContents && !doc.breakContents) {
+        return "";
+      }
+      break;
+  }
```

Comment:
@thorn0 I pushed some changes, I think there is no problem in `concat` part, please review these lines above, are they safe?

Reply:
Please make this function pure. Let's avoid mutating objects when it's not necessary.

Reply:
I read the code again, I think it's pure, I only mutate parts of "concat", I thought it will be faster, and parts already a copy in `mapDoc`.

Reply:
`stripTrailingHardline` was not pure, https://github.com/prettier/prettier/pull/9736/commits/fdbb34cfbdbe80dfc826b19f75d1092a3b16273b

Reply:
I mean `cleanDoc`. But it's better to make `stripTrailingHardline` pure too. Pure functions are cleaner code.

Reply:
Do you mean I should copy `parts` again in `cleanDoc`? Is that really necessary.

Reply:
As for what is faster, I doubt `splice` is faster. Because it needs to shift all the subsequent elements. I wouldn't use `splice` like that without having significant benchmark numbers proving that this really makes sense.